### DISCUSSION TITLE
[🙏 🙏 expert please 🙏 🙏  ]`module.exports` is returned instead of `module.exports.default` when importing from an esm module inside `node_modules` i.e. sub dependency

### DISCUSSION THREAD

**Discussion by bchenSyd** [babel/babel#16112] [Q&A]
hi guys, 

 I'm not too sure if it's a `babel` issue, or `typescript` issue. My question is in the title, but here is a bit more context:

So basically I'm using a 3rd party library which is commonJS module with default export
```js
// XX/index.js
'use strict';
Object.defineProperty(exports, '__esModule', { value: true }); 
exports.default = "this is the export string";
```

when I consume it directly from my project, which is `creat-react-app` (`typescript` + `babel-loader`), below code works fine
```ts
import XX from 'xx';

console.log(XX); // XX is the exported string , i.e. it's    module.exports.default
```

However, I have another ESM library/module `yy`,  which depends on `xx`,and when  I debug the code I noticed that inside `node_modules/yy/index.js`, imported `XX` is the module object , i.e. `module.exports` instead of `module.exports.default`. 
Is is the expect behaviour or is this a defect in `babel`/`typescript`?

```js
// node_module/yy/index.js

import XX from 'xx';

console.log(XX);  // <-------------  XX is an object {__esModule: true, default: 'this is the exported strign'}
export const YY='something'


// src/index.js
import XX from 'xx'; //OK XX is `module.exports.default`
import {YY} from 'yy'; // Error: inside node_module/yy/index.js, XX is `module.exports`

```

can someone please shed some light on this ? thanks! (and sorry I've put the question in the wrong channel)

**Comment by JLHwung** [comment]
> when I consume it directly from my project, which is creat-react-app (typescript + babel-loader), below code works fine

```
import XX from 'xx';

console.log(XX); // XX is the exported string , i.e. it's    module.exports.default
```

It "works" because Babel transforms the above code to

```js
"use strict";

var _xx = _interopRequireDefault(require("xx"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
```
where the helper `_interopRequireDefault` help you unwrap the default exports.

Now when a commonjs module is consumed in Node.js ESM and assuming the node_modules are not handled by Babel, you must import the default export:

```js
import { default as XX } from 'xx'; //OK XX is `module.exports.default`
```

See also https://nodejs.org/api/esm.html#commonjs-namespaces. You may also provide your own `interopRequireDefault` helper.

**Reply by bchenSyd** [reply]
thanks @JLHwung  for answering my question. I totally agree with you and I understand that babel will only do the `interopRequireDefault` for direct import from your project, not `import` inside `node_modules`; 

I'm using `rollup` to build package `yy` and i've marked `xx` as an `external` dependency. So inside `xx`, it will just be  `import YY from 'yy'` , and note that `xx` is an ESM. 

so what's the official solution to this problem? It seems to be quite common case. thanks again

**Comment by bchenSyd** [comment]
> It "works" because Babel transforms the above code to

```js
"use strict";

var _xx = _interopRequireDefault(require("xx"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

```

after some research, I realised that above feature is implemented by `@babel/runtime/helpers/interopRequireDefault`



is there a way to tell babel do the same for `import XX from 'xx'` inside `node_modules`? i.e. if my project depends on `yy`, which is an esm, and `yy` has `import XX from 'xx-cjs'`, can babel return `module.exports.default` for `XX` like it does when you import `xx` directly in your project?


the fact that `babel` returns `module.exports.default` for project import , but `module.exports` for  import  inside `node_modules`, is problematic ( but I can understand the reasoning behind. your babel config should only apply to your project, not packages inside node_modules. as `babel` assume all files inside `node_modules` are already transpiled, so it won't transpile it again) . But how can we transpile an `import`??

**Reply by nicolo-ribaudo** [reply]
> is there a way to tell babel do the same for import XX from 'xx' inside node_modules? i.e. if my project depends on yy, which is an esm, and yy has import XX from 'xx-cjs', can babel return module.exports.default for XX like it does when you import xx directly in your project?

Yes: Babel already does that. Are you sure you are compiling your `node_modules` packages with Babel? How are you running it?

**Reply by bchenSyd** [reply]
hi @nicolo-ribaudo , It seems that you didn't get my question :smile: 

so let's say we have a project that has .babelrc, and it imports a commonjs module `module-A`. here is what that `module-A` looks like
```js
exports.__esModule = true;
exports.default = 'asdf';
```

in your project, if you do `import XX from 'module-A; console.log(XX)`, you see `asdf`  ,  right?





now, let's make it a bit more complicated. Let's say i have another module, call it `module-B`. module-B is a esm.  `module-B` also depends on `module-A`, here is what it looks like
```js
import  XX from 'module-A`; 
console.log(XX)
export const XX;
```

the issue happens when you import `module-A` from your project, all of a sudden, you see `{__esMoulde:true, default: 'asdf'}` in console.log.  Is there a way to let it print `asdf` as well?


make sense?

**Reply by bchenSyd** [reply]
note that this is only an issue when `module-B` is esm; if `module-B` is also commonjs, it would looks like
```js

// 'use strict';

var XX = require('module-A');
function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }
var XX_default = /*#__PURE__*/_interopDefault(XX);
console.log(XX_default.default)
```

and there won't be a problem at all!

**Reply by nicolo-ribaudo** [reply]
> in your project, if you do import XX from 'module-A; console.log(XX), you see asdf , right?

- If you are compiling modules to CommonJS, yes
- If you are running ESM, no you see `{__esMoulde:true, default: 'asdf'}`
- If you are compiling modules to CommonJS with the `importInterop: "node"` option, you see `{__esMoulde:true, default: 'asdf'}`

If you want consistent behavior between native ESM and ESM-compiler-to-CommonJS, you have to enable that option and rewrite your code to
```js
import _A from 'module-A;
const XX = _A.default;

console.log(XX);
```

Unfortunately this is how Node.js decided that importing CommonJS from ESM should work.

When using an ESM module in your `node_modules`, Babel has nothing to do with how `import` works when loading CommonJS: that's entirely Node.js-defined behavior. The most you can do in Babel is to use this option to make sure that Babel behaves like Node.js.

**Reply by bchenSyd** [reply]
perfect! thanks @nicolo-ribaudo ,  we are now on the same page 🎉 

my project is a `webpack` react project and I'm using `babel-loader`, so yes the **project code** will be transpiled by `babel`. However, `module-A` code will keep the node.js behavior since it lives in `node_modules`.
this is where the **inconsistency** comes from;


I'm using `rollup` to build `modal-B`, and declare `modal-A` as an `external dependency` there; 
here is my `modal-B` compiled code (with `@rollup/plugin-babel`)
```js
// modal-B is compiled using rollup as esm output
import AA from 'modal-A';

export const BB = AA;
```
I'm not sure how to instruct `rollup` to add `export const B= AA.default`. However,  if I set `rollup` output format to `cjs`, it injects below code
```js
function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }
var AA__default = /*#__PURE__*/_interopDefault(AA);
```
so it's all good in this scenario.  The inconsistency only happens when `rollup` outputs format `esm` 

 hope it's clear. thanks @nicolo-ribaudo  again!

**Reply by bchenSyd** [reply]
> When using an ESM module in your node_modules, Babel has nothing to do with how import works when loading CommonJS: that's entirely Node.js-defined behavior. The most you can do in Babel is to use this option to make sure that Babel behaves like Node.js.


I can't agree more on what you said

**Reply by bchenSyd** [reply]
@nicolo-ribaudo  i've raised an [issue](https://github.com/rollup/rollup/issues/5265) in `rollup`. feel free to add comments there

**Comment by bchenSyd** [comment]
@JLHwung

**Comment by nicolo-ribaudo** [comment]
A possible solution is to pass the `importInterop: "node"` option (https://babeljs.io/docs/babel-plugin-transform-modules-commonjs#importinterop) to Babel, so that it compiles your non-node_modules files the same way Node.js would run them (i.e. without `_interopRequireDefault`).
